### PR TITLE
ArrayData Enumeration for Remaining Layouts

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-rs/issues/1799

# Rationale for this change
 

Follow on to https://github.com/apache/arrow-rs/pull/3749 adds more types

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-data/src/data/bytes.rs [hunk 1]
```diff
@@ -270,7 +270,7 @@ impl<O: BytesOffset, B: Bytes> BytesArrayData<O, B> {
 
     /// Returns the offsets
     #[inline]
-    pub fn value_offsets(&self) -> &[O] {
+    pub fn offsets(&self) -> &[O] {
```
Comment (tustvold) [inline arrow-data/src/data/bytes.rs:273]: The value_ prefix is redundant

File: arrow-data/src/data/primitive.rs [hunk 2]
```diff
@@ -43,13 +43,13 @@ mod private {
 }
 
 pub trait Primitive: private::PrimitiveSealed + ArrowNativeType {
-    const VARIANT: PrimitiveType;
+    const TYPE: PrimitiveType;
```
Comment (tustvold) [inline arrow-data/src/data/primitive.rs:46]: This makes it consistent with the other traits

File: arrow-data/src/data/types.rs [hunk 3]
```diff
@@ -141,7 +140,7 @@ impl From<&DataType> for PhysicalType {
                 DataType::UInt64 => Self::Dictionary(DictionaryKeyType::UInt64),
                 d => panic!("illegal dictionary key data type {d}"),
             },
-            DataType::Map(_, _) => Self::Map,
+            DataType::Map(_, _) => Self::List(OffsetType::Int32),
```
Comment (tustvold) [inline arrow-data/src/data/types.rs:143]: MapArray is just a logical type over ListArray

File: arrow-data/src/data/run.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+impl<E: RunEnd> From<RunArrayData<E>> for ArrayDataRun {
+    fn from(value: RunArrayData<E>) -> Self {
+        E::upcast(value)
+    }
+}
+
+/// ArrayData for [run-end encoded arrays](https://arrow.apache.org/docs/format/Columnar.html#run-end-encoded-layout)
+pub struct RunArrayData<E: RunEnd> {
```
Comment (tustvold) [inline arrow-data/src/data/run.rs:107]: This is a slight deviation from the spec as defined in https://arrow.apache.org/docs/format/Columnar.html#run-end-encoded-layout in that it "inlines" the run ends. This is consistent with `RunEndArray` and I think makes for a significantly less confusing abstraction.

**Review by alamb — APPROVED** [review]
The patterns all look good to me. Thanks @tustvold

**Comment by tustvold** [comment]
Unless I hear anything further, I plan to merge this tomorrow morning. Please let me know if you need more time to review

**Comment by viirya** [comment]
I will take some time looking at this today but if I cannot, it don't need to block it.

File: arrow-data/src/data/dictionary.rs [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            Self: DictionaryKey;
+
+        /// Cast [`DictionaryArrayData`] to [`ArrayDataDictionary`]
+        fn upcast(v: DictionaryArrayData<Self>) -> ArrayDataDictionary
+        where
+            Self: DictionaryKey;
+    }
+}
+
+/// Types of offset used by variable length byte arrays
```
Comment (viirya) [inline arrow-data/src/data/dictionary.rs:45]: Seems wrong due to copying?

File: arrow-data/src/data/list.rs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+            Self: ListOffset;
+
+        /// Cast [`ListArrayData`] to [`ArrayDataList`]
+        fn upcast(v: ListArrayData<Self>) -> ArrayDataList
+        where
+            Self: ListOffset;
+    }
+}
+
+/// Types of offset used by variable length byte arrays
```
Comment (viirya) [inline arrow-data/src/data/list.rs:45]: ditto.

File: arrow-data/src/data/list.rs [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    child: Box<ArrayData>,
+}
+
+impl<O: ListOffset> ListArrayData<O> {
+    /// Create a new [`ListArrayData`]
+    ///
+    /// # Safety
+    ///
+    /// - Each consecutive window of `offsets` must identify a valid slice of `child`
+    /// - `nulls.len() == offsets.len() + 1`
```
Comment (viirya) [inline arrow-data/src/data/list.rs:156]:
```suggestion
    /// - `nulls.len() == offsets.len() - 1`
```
?

File: arrow-data/src/data/union.rs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    ) -> Self {
+        Self {
+            data_type,
+            type_ids,
+            offsets,
+            children,
+        }
+    }
+
+    /// Returns the type ids for this array if this is a dense union
```
Comment (viirya) [inline arrow-data/src/data/union.rs:54]: sparse union also has type ids.
```suggestion
    /// Returns the type ids for this array
```

**Review by viirya — APPROVED** [review]

**Pull request merged**